### PR TITLE
Add compatibility with Tumblr API links objects being passed through to these low level functions

### PR DESCRIPTION
**Pull request by cyle**
See https://github.com/tumblr/tumblr.js/issues/99 for some backstory.

tl;dr you should be able to do this cleanly: `tumblr.getRequest(response._links.next.href);` when `response._links.next` is one of the links objects from [our API](https://github.com/tumblr/docs/blob/master/api.md#links).

This is a tricky one (though that's not really obvious) for a few reasons:

1. The `tumblr.js` library sets the base URL to include `/v2`: https://github.com/tumblr/tumblr.js/blob/master/lib/tumblr.js#L30
2. The `tumblr.js` library also assumes that the given `apiPath` never has existing query params, which is true for the helper methods we're including, but not necessarily true if you want to dive all the way down to `tumblr.getRequest` itself.
3. In the Tumblr backend, we do not assume the client is doing either of these things, so we supply the full URL path and query string parameters for `GET` requests.

We can't reasonably change the third point on the backend; I'd rather the backend _not_ change its behavior based on the client, and instead fix this at the `tumblr.js` framework level, since we're baking assumptions in there for its specific use case. That feels like the right move here but I'm open to other possibilities. I flirted with the idea of adding a `path` field in the backend for all of these `links` objects, but this would bloat... just about every response size we serve, when it's fairly standard to have a means on the consumer side to parse that `href`, as I'm doing here.

Thoughts? @shahkashani @sndsgd

### CODE REVIEW

File: lib/tumblr.js [hunk 1]
```diff
@@ -478,6 +478,23 @@ function getRequest(requestGet, credentials, baseUrl, apiPath, requestOptions, p
         params.api_key = credentials.consumer_key;
     }
 
+    // if given /v2/user/info, remove the /v2
+    if (apiPath.substr(0, 3) === '/v2') {
```
Comment (shahkashani) [inline lib/tumblr.js:482]: Not a big deal, but could probably use [`startsWith`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith) here.

Comment (cyle) [reply lib/tumblr.js:482]: ah yeah, i am so far behind what things i should expect to be available to me in javascript 😅

Comment (shahkashani) [reply lib/tumblr.js:482]: It's generally safe to assume nothing is free in JavaScript 😆

Comment (cyle) [reply lib/tumblr.js:482]: in this case i think i'll leave it as-is as to not drop IE support, i don't wanna cause that headache here

Comment (shahkashani) [reply lib/tumblr.js:482]: Do you wanna refactor this into a method since you're already doing it twice here? Something like `getApiPath = (path) => path.substr(0, 3) === '/v2' ? path.substr(3) : path`

**Comment by sndsgd** [comment]
is there a reason we cannot update all the routes in `API_METHODS` to be the full path and update `API_BASE_URL` to not include a path?

```patch
diff --git a/lib/tumblr.js b/lib/tumblr.js
index 40dcd2f..f790fc3 100644
--- a/lib/tumblr.js
+++ b/lib/tumblr.js
@@ -27,7 +27,7 @@ const isPlainObject = require('lodash/isPlainObject');
 const omit = require('lodash/omit');
 
 const CLIENT_VERSION = '1.0.0';
-const API_BASE_URL = 'https://api.tumblr.com/v2';
+const API_BASE_URL = 'https://api.tumblr.com';
 
 const API_METHODS = {
     GET: {
@@ -44,7 +44,7 @@ const API_METHODS = {
          *
          * @memberof TumblrClient
          */
-        blogInfo: '/blog/:blogIdentifier/info',
+        blogInfo: '/v2/blog/:blogIdentifier/info',
 
         /**
          * Gets the avatar URL for a blog
```

File: lib/tumblr.js (hunk 1 above)
Comment (nightpool) [reply lib/tumblr.js:482]: I can't imagine this library ever gets used on IE, since it requires node libraries and utils? So the relevant version would be 0.12 behind the harmony flag or 4.0.0 without the flag

Comment (nightpool) [reply lib/tumblr.js:482]: (so imo we're well clear to go ahead and use startsWith)

Comment (shahkashani) [reply lib/tumblr.js:482]: You're totally right. None of this stuff works in a browser anyway.

Comment (cyle) [reply lib/tumblr.js:482]: oh, lol, yeah, good points

**Comment by cyle** [comment]
> is there a reason we cannot update all the routes in API_METHODS to be the full path and update API_BASE_URL to not include a path?

I guess not, since we control that. It'd only be tricky if someone has decided to change the `baseUrl` they pass in when instantiating the base `TumblrClient`, this change would break it. But then again that's why we have semantic versioning. That would clear up one part of this.

**Comment by cyle** [comment]
Updated to change the base URL to exclude `/v2/` and endpoint paths to include `/v2/`, updated version from 2.0.2 to 3.0.0.

File: lib/tumblr.js [hunk 2]
```diff
@@ -26,8 +26,8 @@ const isArray = require('lodash/isArray');
 const isPlainObject = require('lodash/isPlainObject');
 const omit = require('lodash/omit');
 
-const CLIENT_VERSION = '1.0.0';
-const API_BASE_URL = 'https://api.tumblr.com/v2';
+const CLIENT_VERSION = '3.0.0';
```
Comment (cyle) [inline lib/tumblr.js:29]: looks like this never got updated with the actual version in the package.json file :eyes:

**Comment by nightpool** [comment]
@cyle @sndsgd one other consideration here is that changing BASE_PATH would break any other code currently using getRequest/postRequest, which seems like a _much_ more annoying and invasive change for calling code to have to make, depending on how much structure they have around getRequest entirely. 

I'm generally in favor of the semantic versioning of it all, and if we ever wanted to do a `v3` route then we'd need _some_ way of handling it, but just wanted to flag that it's a little bit more complicated then just "anybody who's decided to change the baseUrl"

**Comment by cyle** [comment]
Yeah that's valid @nightpool . I'm still in favor of @sndsgd 's suggestion, if only because I hate having to do arbitrary string juggling in our own framework to support what we are doing on the backend for the rest of our official clients. I'd rather fit this framework better into how things work from the backend, which should help prevent future issues like this. I see this as adjusting the "happy path", even though that means a more invasive change for anyone who has been playing this deep in the plumbing. Ideally, seeing that it's a major version bump should mean that they'll register that they may need to fix up that plumbing. 🤞 

Thoughts @shahkashani @sndsgd ?

**Comment by nightpool** [comment]
aligning with server-side semantics is a compelling enough argument for me! 👍

**Comment by cyle** [comment]
That sounds great to me. I love friendly and helpful error messages!

**Comment by cyle** [comment]
Ok, I think this has got to a meaningfully good place. Thoughts @sndsgd @shahkashani @nightpool ?
